rfcs/0028-dag-structured_query_plan.md

```diff
@@ -0,0 +1,67 @@
+
+### Share Operator
+
+In order to represent the DAG plan during the whole optimization time, we can use a `Share` operator to explicitly represent the sub-plan having multiple parents. Why do we need to add an additional operator instead of directly letting its parent point to the sub-plan? Because we have lots of optimizer transformations which will clone and modify the subquery and finally break the DAG plan back to a tree plan. Using a `Share` operator we can tell the optimizer we want to keep the sub-plan unchanged during the optimization.
+
+![](images/share_operator.svg)
+
+The `Share` operator contains multiple parents. Each of them wants to push the predicate and prune the columns of it, so we need to merge them instead of directly affecting the sub-plan. We can create a new optimization stage for merging all the predicates pushed down from the parent of the `Share` operator. The merge operation for filters is to `OR` them together. We can merge the required columns from the parents in a similar way. The merge operation for the column pruning is to `UNION` them together.
+
+
+#### Example
+With the `Share` operator we can perform common plan detection at any time of the optimization. I will give some examples:
+
+- CTE. At the binder/planner stage, we can create a `Share` for the table expression and reuse it.
+- Subquery unnesting. We can calculate the domain by creating a `Share` operator of the LHS of the apply operator.
+- Same source operator appears multiple times in a query. Use Share operator at binder/planner or at end of optimizer are both ok.
+- Common sub-plan detection. We can use a field `digest` for each sub-plan and if two sub-plan have the same `digest` it means they are probably the same sub-plan. `digest` is just like a hashcode. We also need to have a `deepEqual` method for each plan to compare with another one.
+
+![](images/self_join.svg)
+
+![](images/subquery_unnesting.svg)
+
+Last but not least, at the end of the optimization, we might find that the `Share` operator has only one parent. In this case, we can remove the `Share` operator.
+
+#### Cost
+
+The cost of DAG structured plans can be complex if we have a large search space, especially for batch query. Luckily for streaming queries, we don't have a large search space for now and we can always convert the plan to the DAG plan if possible or provide a configuration for users to determine whether to use it or not.
+
+For the batch query, DAG means we need to read the input multiple times which inevitably leads to buffer or materialization. The cost of DAG is determined by how many times it has been read. Normally, the first time has a higher cost than later ones, because of the materialization. It is hard to tell whether to use DAG or not for batch query, I think we can always convert the DAG back to the tree for batch executor.
+
+For the streaming query, it seems not a big deal. We support buffering and dispatch naturally.
```

Comment:
Interesting observation. Indeed, for streaming, execution of DAG plan doesn't introduce any additional cost. No extra buffering needed.